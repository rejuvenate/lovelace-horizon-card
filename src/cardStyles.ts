import { css } from 'lit'

export default css`
  .horizon-card {
    --horizon-card-primary: var(--primary-text-color, #000000);
    --horizon-card-secondary: var(--secondary-text-color, #828282);
    --horizon-card-accent: #d7d7d7;

    --horizon-card-lines: var(--horizon-card-accent);
    --horizon-card-field-name-color: var(--horizon-card-secondary);
    --horizon-card-field-value-color: var(--horizon-card-primary);

    --horizon-card-stop-invisible: rgb(0,0,0,0);
    --horizon-card-stop-sun-color: #f9d05e;
    --horizon-card-stop-dawn-color: #393b78;
    --horizon-card-stop-day-color: #8ebeeb;
    --horizon-card-stop-dusk-color: #393b78;

    padding: 0.5rem;
    font-size: 1.3rem;
    font-family: var(--primary-font-family);
  }

  .horizon-card.horizon-card-dark {
    --horizon-card-primary: #ffffff;
    --horizon-card-secondary: #828282;
    --horizon-card-accent: #464646;
  }

  .horizon-card-field-row {
    display: flex;
    justify-content: space-around;
    margin: 1rem 1.5rem 0 1.5rem;
  }

  .horizon-card-text-container {
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .horizon-card-field-name {
    color: var(--horizon-card-field-name-color);
  }

  .horizon-card-field-value {
    color: var(--horizon-card-field-value-color);
  }

  .horizon-card-header {
    display: flex;
    justify-content: space-between;
    margin: 0 4rem 0 4rem;
  }

  .horizon-card-header .horizon-card-field-value {
    font-size: 1.85rem;
  }

  .horizon-card-title {
    margin: 0.5rem 0.5rem 3rem 3rem;
    font-size: 2rem;
    font-weight: 500;
    color: var(--horizon-card-primary);
  }

  .horizon-card-graph {
    shape-rendering="geometricPrecision";
    margin: 2rem 0 2rem 0;
  }

  .horizon-card-graph .sunInitialStop {
    stop-color: var(--horizon-card-stop-sun-color);
  }

  .horizon-card-graph .sunMiddleStop {
    stop-color: var(--horizon-card-stop-sun-color);
  }

  .horizon-card-graph .sunEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .horizon-card-graph .dawnInitialStop {
    stop-color: var(--horizon-card-stop-dawn-color);
  }

  .horizon-card-graph .dawnMiddleStop {
    stop-color: var(--horizon-card-stop-dawn-color);
  }

  .horizon-card-graph .dawnEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .horizon-card-graph .dayInitialStop {
    stop-color: var(--horizon-card-stop-day-color);
  }

  .horizon-card-graph .dayMiddleStop {
    stop-color: var(--horizon-card-stop-day-color);
  }

  .horizon-card-graph .dayEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .horizon-card-graph .duskInitialStop {
    stop-color: var(--horizon-card-stop-dusk-color);
  }

  .horizon-card-graph .duskMiddleStop {
    stop-color: var(--horizon-card-stop-dusk-color);
  }

  .horizon-card-graph .duskEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .card-config ul {
    list-style: none;
    padding: 0 0 0 1.5rem;
  }

  .card-config li {
    padding: 0.5rem 0;
  }
`
