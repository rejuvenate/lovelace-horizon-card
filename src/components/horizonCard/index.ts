export * from './HorizonCard'
export * from './HorizonCardContent'
export * from './HorizonCardFooter'
export * from './HorizonCardGraph'
export * from './HorizonCardHeader'
