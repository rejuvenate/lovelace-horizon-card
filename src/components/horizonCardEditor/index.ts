export * from './HorizonCardEditor'
export * from './HorizonCardEditorContent'
