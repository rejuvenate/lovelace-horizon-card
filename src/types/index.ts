import { LovelaceCardConfig } from 'custom-card-helpers'

import { I18N } from '../utils/I18N'

export type THorizonCardFields = {
  sunrise?: boolean
  sunset?: boolean

  dawn?: boolean
  noon?: boolean
  dusk?: boolean

  azimuth?: boolean
  elevation?: boolean
}

export interface IHorizonCardConfig extends LovelaceCardConfig {
  i18n?: I18N
  darkMode?: boolean
  language?: string

  title?: string,
  component?: string
  use12hourClock?: boolean

  fields?: THorizonCardFields
}

export interface IConfigChangedEvent <T> extends CustomEvent<T> {
  target: CustomEvent['target'] & {
    configValue: string
    selected?: string
    checked?: boolean
  }
}

export type TSunInfo = {
  sunrise: number,
  sunset: number

  dawnProgressPercent: number
  dayProgressPercent: number
  duskProgressPercent: number

  sunAboveHorizon: boolean
  sunPercentOverHorizon: number
  sunPosition: {
    x: number
    y: number
  }
}

export enum EHorizonCardErrors {
  SunIntegrationNotFound = 'SunIntegrationNotFound'
}

export type THorizonCardTimes = {
  dawn?: Date
  dusk?: Date
  noon: Date
  sunrise?: Date
  sunset?: Date
}

export type THorizonCardData = {
  azimuth: number
  elevation: number

  sunInfo: TSunInfo

  times: THorizonCardTimes

  error?: EHorizonCardErrors
}

export enum EHorizonCardI18NKeys {
  Azimuth = 'azimuth',
  Dawn = 'dawn',
  Dusk = 'dusk',
  Elevation = 'elevation',
  Noon = 'noon',
  Sunrise = 'sunrise',
  Sunset = 'sunset'
}

export type THorizonCardI18N = Record<string, unknown>

export type THorizonCardI18NErrorKeys = {
  [key in EHorizonCardErrors]: string
}

export type THorizonCardI18NKeys = { [key in EHorizonCardI18NKeys ]: string } | { errors: THorizonCardI18NErrorKeys }
