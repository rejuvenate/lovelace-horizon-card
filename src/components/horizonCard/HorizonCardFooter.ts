import { html, TemplateResult } from 'lit'

import { EHorizonCardI18NKeys, IHorizonCardConfig, THorizonCardData, THorizonCardFields, THorizonCardTimes } from '../../types'
import { HelperFunctions } from '../../utils/HelperFunctions'
import { I18N } from '../../utils/I18N'

export class HorizonCardFooter {
  private data: THorizonCardData
  private i18n: I18N
  private times: THorizonCardTimes
  private fields: THorizonCardFields

  constructor (config: IHorizonCardConfig, data: THorizonCardData) {
    this.data = data

    this.i18n = config.i18n!
    this.times = data?.times
    this.fields = config.fields!
  }

  public render (): TemplateResult {
    return html`
      <div class="horizon-card-footer">
        <div class="horizon-card-field-row">
          ${
  this.fields?.dawn !== undefined
    ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Dawn, this.times?.dawn)
    : HelperFunctions.nothing()
}
          ${
  this.fields?.noon !== undefined
    ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Noon, this.times?.noon)
    : HelperFunctions.nothing()
}
          ${
  this.fields?.dusk !== undefined
    ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Dusk, this.times?.dusk)
    : HelperFunctions.nothing()
}
        </div>

        <div class="horizon-card-field-row">
          ${
  this.fields?.azimuth !== undefined
    ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Azimuth, this.data?.azimuth)
    : HelperFunctions.nothing()
}
          ${
  this.fields?.elevation !== undefined
    ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Elevation, this.data?.elevation)
    : HelperFunctions.nothing()
}
        </div>
      </div>
    `
  }
}
