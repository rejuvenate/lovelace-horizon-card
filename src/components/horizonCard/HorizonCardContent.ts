import { html,TemplateResult } from 'lit'

import { IHorizonCardConfig,THorizonCardData } from '../../types'
import { HelperFunctions } from '../../utils/HelperFunctions'
import { HorizonCardFooter } from './HorizonCardFooter'
import { HorizonCardGraph } from './HorizonCardGraph'
import { HorizonCardHeader } from './HorizonCardHeader'

export class HorizonCardContent {
  private config: IHorizonCardConfig
  private data: THorizonCardData

  constructor (config: IHorizonCardConfig, data: THorizonCardData) {
    this.config = config
    this.data = data
  }

  render (): TemplateResult {
    return html`
      <ha-card>
        <div class="horizon-card ${this.config.darkMode ? 'horizon-card-dark' : ''}">
          ${ this.showHeader() ? this.renderHeader() : HelperFunctions.nothing() }
          ${ this.renderGraph() }
          ${ this.showFooter() ? this.renderFooter() : HelperFunctions.nothing() }
        </div>
      </ha-card>
    `
  }

  private renderHeader (): TemplateResult {
    return new HorizonCardHeader(this.config, this.data).render()
  }

  private renderGraph (): TemplateResult {
    return new HorizonCardGraph(this.data).render()
  }

  private renderFooter (): TemplateResult {
    return new HorizonCardFooter(this.config, this.data).render()
  }

  private showHeader (): boolean {
    // logic based on config
    return true
  }

  private showFooter (): boolean {
    // logic based on config
    return true
  }
}
