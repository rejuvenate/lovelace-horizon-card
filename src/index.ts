import './components/horizonCard'
import './components/horizonCardEditor'

