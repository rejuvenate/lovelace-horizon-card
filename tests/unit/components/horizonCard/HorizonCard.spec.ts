import { HomeAssistant } from 'custom-card-helpers'
import { css, CSSResult } from 'lit'

import { HorizonCard } from '../../../../src/components/horizonCard'
import { HorizonCardEditor } from '../../../../src/components/horizonCardEditor'
import { Constants } from '../../../../src/constants'
import { EHorizonCardErrors, IHorizonCardConfig, THorizonCardData } from '../../../../src/types'
import { CustomSnapshotSerializer, TemplateResultTestHelper } from '../../../helpers/TestHelpers'

jest.mock('../../../../src/components/HorizonErrorContent', () => require('../../../mocks/HorizonErrorContent'))
jest.mock('../../../../src/components/horizonCard/HorizonCardContent', () => require('../../../mocks/HorizonCardContent'))
jest.mock('../../../../src/utils/HelperFunctions', () => require('../../../mocks/HelperFunctions'))
jest.mock('../../../../src/utils/I18N', () => require('../../../mocks/I18N'))
jest.mock('../../../../src/cardStyles', () => css``)

expect.addSnapshotSerializer(new CustomSnapshotSerializer())

// JSDom doesn't include SVGPathElement
class SVGPathElement {
  getPointAtLength () {
    return {
      x: 0,
      y: 0
    }
  }
}

Object.defineProperty(window, 'SVGPathElement', { value: SVGPathElement })

describe('HorizonCard', () => {
  let horizonCard: HorizonCard

  beforeEach(() => {
    horizonCard = new HorizonCard()
    horizonCard.attachShadow({ mode: 'open' })
  })

  describe('set hass', () => {
    it('updates lastHass property', () => {
      jest.spyOn(horizonCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      jest.spyOn(horizonCard as any, 'processLastHass').mockReturnValue(undefined)
      expect(horizonCard['lastHass']).toBeUndefined()

      const expectedLastHass = {} as HomeAssistant
      horizonCard.hass = expectedLastHass
      expect(horizonCard['lastHass']).toEqual(expectedLastHass)
    })

    it('calls populateConfigFromHass if it has not been rendered yet', () => {
      const populateConfigFromHassSpy = jest.spyOn(horizonCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      jest.spyOn(horizonCard as any, 'processLastHass').mockReturnValue(undefined)
      horizonCard['hasRendered'] = false

      horizonCard.hass = {} as HomeAssistant
      expect(populateConfigFromHassSpy).toHaveBeenCalledTimes(1)
    })

    it('does not call populateConfigFromHass if it has been rendered', () => {
      const populateConfigFromHassSpy = jest.spyOn(horizonCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      jest.spyOn(horizonCard as any, 'processLastHass').mockReturnValue(undefined)
      horizonCard['hasRendered'] = true

      horizonCard.hass = {} as HomeAssistant
      expect(populateConfigFromHassSpy).not.toHaveBeenCalled()
    })

    it('calls processLastHass if it has been rendered', () => {
      jest.spyOn(horizonCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      const processLastHassSpy = jest.spyOn(horizonCard as any, 'processLastHass').mockReturnValue(undefined)
      horizonCard['hasRendered'] = true

      horizonCard.hass = {} as HomeAssistant
      expect(processLastHassSpy).toHaveBeenCalledTimes(1)
    })
  })

  describe('getConfigElement', () => {
    it('creates and return a horizon card config element', () => {
      const expectedElement = document.createElement('div')
      const createElementSpy = jest.spyOn(document, 'createElement').mockReturnValueOnce(expectedElement)
      const result = HorizonCard.getConfigElement()

      expect(result).toEqual(expectedElement)
      expect(createElementSpy).toHaveBeenCalledTimes(1)
      expect(createElementSpy).toHaveBeenCalledWith(HorizonCardEditor.cardType)
    })
  })

  describe('setConfig', () => {
    it('overrides old config with new values', () => {
      jest.spyOn(horizonCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      expect(horizonCard['config']).toEqual({ type: HorizonCard.cardType })
      const config = {
        type: HorizonCard.cardType,
        title: 'Test'
      } as IHorizonCardConfig

      horizonCard.setConfig(config)
      expect(horizonCard['config'].title).toEqual(config.title)

      const newConfig = {
        type: HorizonCard.cardType,
        title: 'Test2'
      } as IHorizonCardConfig

      horizonCard.setConfig(newConfig)
      expect(horizonCard['config'].title).toEqual(newConfig.title)
    })

    it('uses the provided component', () => {
      const config = {
        type: HorizonCard.cardType,
        component: 'test'
      } as IHorizonCardConfig

      horizonCard.setConfig(config)
      expect(horizonCard['config'].component).toEqual(config.component)
    })

    it('uses the default component when not provided', () => {
      const config = {
        type: HorizonCard.cardType
      } as IHorizonCardConfig

      horizonCard.setConfig(config)
      expect(horizonCard['config'].component).toEqual(Constants.DEFAULT_CONFIG.component)
    })

    it('throws an error if the provided language is not supported', () => {
      const config = {
        type: HorizonCard.cardType,
        language: 'notSupportedLanguage'
      } as IHorizonCardConfig

      let thrownError
      try {
        horizonCard.setConfig(config)
      } catch (error) {
        thrownError = (error as Error).message
      }

      expect(thrownError).toEqual(`${config.language} is not a supported language. Supported languages: ${Object.keys(Constants.LOCALIZATION_LANGUAGES)}`)
    })

    const fields = ['sunrise', 'sunset', 'dawn', 'noon', 'dusk', 'azimuth', 'elevation']
    for (const field of fields) {
      it(`uses the provided value for ${field}`, () => {
        const config = {
          type: HorizonCard.cardType,
          fields: {
            [field]: 'test'
          }
        } as IHorizonCardConfig

        horizonCard.setConfig(config)
        expect(horizonCard['config'].fields?.[field]).toEqual('test')
      })

      it(`uses the default value for ${field} when not provided`, () => {
        const config = {
          type: HorizonCard.cardType
        } as IHorizonCardConfig

        horizonCard.setConfig(config)
        expect(horizonCard['config'].fields?.[field]).toEqual(Constants.DEFAULT_CONFIG.fields?.[field])
      })
    }

    it('calls populateConfigFromHass if lastHass has a value', () => {
      const populateConfigFromHassSpy = jest.spyOn(horizonCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      horizonCard['lastHass'] = {} as HomeAssistant

      const config = {
        type: HorizonCard.cardType
      } as IHorizonCardConfig

      horizonCard.setConfig(config)
      expect(populateConfigFromHassSpy).toHaveBeenCalledTimes(1)
    })

    it('does not call populateConfigFromHass if lastHass has not a value', () => {
      const populateConfigFromHassSpy = jest.spyOn(horizonCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      delete (horizonCard as any)['lastHass']

      const config = {
        type: HorizonCard.cardType
      } as IHorizonCardConfig

      horizonCard.setConfig(config)
      expect(populateConfigFromHassSpy).not.toHaveBeenCalled()
    })
  })

  describe('render', () => {
    it('renders an error if error is present on data', async () => {
      if (!horizonCard['data']) {
        horizonCard['data'] = {} as THorizonCardData
      }

      horizonCard['data'].error = EHorizonCardErrors.SunIntegrationNotFound
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof horizonCard.render>
      element.templateResultFunction = () => horizonCard.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('renders the horizon card if no error is present on data', async () => {
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof horizonCard.render>
      element.templateResultFunction = () => horizonCard.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })
  })

  describe('updated', () => {
    it('sets to true hasRendered if has not been rendered yet', () => {
      jest.spyOn(horizonCard as any, 'processLastHass').mockReturnValue(undefined)
      horizonCard['hasRendered'] = false
      horizonCard['updated'](new Map())
      expect(horizonCard['hasRendered']).toEqual(true)
    })

    it('calls processLastHass if has not been rendered yet', () => {
      const processLastHassSpy = jest.spyOn(horizonCard as any, 'processLastHass').mockReturnValue(undefined)
      horizonCard['hasRendered'] = false
      horizonCard['updated'](new Map())
      expect(processLastHassSpy).toHaveBeenCalledTimes(1)
    })

    it('does nothing more than call super.updated if has been already rendered', () => {
      const processLastHassSpy = jest.spyOn(horizonCard as any, 'processLastHass').mockReturnValue(undefined)
      horizonCard['hasRendered'] = true
      horizonCard['updated'](new Map())
      expect(processLastHassSpy).not.toHaveBeenCalledWith()
    })
  })

  describe('populateConfigFromHass', () => {
    it('keeps values from config when present', () => {
      const config = {
        type: HorizonCard.cardType,
        darkMode: true,
        language: 'es'
      } as IHorizonCardConfig

      horizonCard['lastHass'] = {
        themes: {
          darkMode: false
        },
        locale: {
          language: 'en'
        }
      } as any

      horizonCard['config'] = config
      horizonCard['populateConfigFromHass']()
      expect(horizonCard['config'].darkMode).toEqual(config.darkMode)
      expect(horizonCard['config'].language).toEqual(config.language)
    })

    it('sets values from hass when not present in config', () => {
      horizonCard['lastHass'] = {
        themes: {
          darkMode: false
        },
        locale: {
          language: 'it'
        }
      } as any

      horizonCard['populateConfigFromHass']()
      expect(horizonCard['config'].darkMode).toEqual(false)
      expect(horizonCard['config'].language).toEqual('it')
    })

    it('supports old versions of home assistant', () => {
      horizonCard['lastHass'] = {
        themes: {
          darkMode: false
        },
        language: 'it'
      } as any

      horizonCard['populateConfigFromHass']()
      expect(horizonCard['config'].darkMode).toEqual(false)
      expect(horizonCard['config'].language).toEqual('it')
    })
  })

  describe('processLastHass', () => {
    it('does an early return if lastHass has no value', () => {
      delete (horizonCard as any)['lastHass']
      const populateConfigFromHassSpy = jest.spyOn(horizonCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      horizonCard['processLastHass']()

      expect(populateConfigFromHassSpy).not.toHaveBeenCalled()
    })

    it('calls populateConfigFromHass', () => {
      (horizonCard as any)['lastHass'] = { states: {} }
      const populateConfigFromHassSpy = jest.spyOn(horizonCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      horizonCard['processLastHass']()

      expect(populateConfigFromHassSpy).toHaveBeenCalledTimes(1)
    })

    it('process the data from the component when found', () => {
      (horizonCard as any)['lastHass'] = {
        states: {
          component: {
            attributes: {
              azimuth: 3,
              elevation: 7
            }
          }
        }
      }

      const config = {
        type: HorizonCard.cardType,
        component: 'component'
      } as IHorizonCardConfig

      horizonCard['config'] = config

      jest.spyOn(horizonCard as any, 'populateConfigFromHass').mockReturnValue(undefined)

      const times = {
        dawn: new Date(0),
        dusk: new Date(0),
        noon: new Date(0),
        sunrise: new Date(0),
        sunset: new Date(0)
      }

      const sunInfo = {
        sunrise: 0,
        sunset: 0,
        dawnProgressPercent: 0,
        dayProgressPercent: 0,
        duskProgressPercent: 0,
        sunAboveHorizon: 0,
        sunPercentOverHorizon: 0,
        sunPosition: { x: 0, y: 0 }
      }

      const readTimesSpy = jest.spyOn(horizonCard as any, 'readTimes').mockReturnValue(times)
      const calculateSunInfoSpy = jest.spyOn(horizonCard as any, 'calculateSunInfo').mockReturnValue(sunInfo)

      horizonCard['processLastHass']()

      expect(readTimesSpy).toHaveBeenCalledTimes(1)
      expect(calculateSunInfoSpy).toHaveBeenCalledTimes(1)
      expect(horizonCard['data'].times).toEqual(times)
      expect(horizonCard['data'].sunInfo).toEqual(sunInfo)
    })

    it('sets an error if the component is not found', () => {
      (horizonCard as any)['lastHass'] = { states: {} }
      jest.spyOn(horizonCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      horizonCard['processLastHass']()

      expect(horizonCard['data'].error).toEqual(EHorizonCardErrors.SunIntegrationNotFound)
    })
  })

  describe('readTimes', () => {
    it('returns dates for each time field', () => {
      const readTimeSpy = jest.spyOn(horizonCard as any, 'readTime').mockReturnValue(new Date(0))
      const result = horizonCard['readTimes']({
        next_setting: 0,
        next_dawn: 0,
        next_dusk: 0,
        next_noon: 0,
        next_rising: 0
      })

      expect(result).toEqual({
        dawn: new Date(0),
        dusk: new Date(0),
        noon: new Date(0),
        sunrise: new Date(0),
        sunset: new Date(0)
      })

      expect(readTimeSpy).toHaveBeenCalledTimes(4)
    })
  })

  describe('readTime', () => {
    it('sets a specific day, month and year to a provided string date', () => {
      // Prepare
      const attributeToParse = '2023-03-18T00:00:00.000Z'
      const year = 2023
      const month = 1
      const date = 1

      // Act
      const result = horizonCard['readTime'](attributeToParse, year, month, date)

      // Assert
      expect(result).toBeInstanceOf(Date)
      expect(result.getUTCFullYear()).toEqual(year)
      expect(result.getUTCMonth()).toEqual(month)
      expect(result.getUTCDate()).toEqual(date)
    })
  })

  describe('calculateSunInfo', () => {
    it('returns all sun info', () => {
      const path = new SVGPathElement()
      jest.spyOn((horizonCard as any).shadowRoot, 'querySelector').mockReturnValue(path)

      const result = horizonCard['calculateSunInfo'](new Date(0), new Date(0))
      expect(result).toEqual({
        dawnProgressPercent: 0,
        dayProgressPercent: 0,
        duskProgressPercent: 0,
        sunAboveHorizon: true,
        sunPercentOverHorizon: 0,
        sunPosition: {
          x: 0,
          y: 0
        },
        sunrise: 0,
        sunset: 0
      })
    })
  })

  describe('get styles', () => {
    it('returns a CSSResult', () => {
      expect(HorizonCard.styles).toBeInstanceOf(CSSResult)
    })
  })
})
