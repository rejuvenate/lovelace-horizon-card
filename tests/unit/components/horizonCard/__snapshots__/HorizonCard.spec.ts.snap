// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizonCard render renders an error if error is present on data 1`] = `
<!---->
      <div class="horizon-card-error">
        ERROR
      </div>
    
`;

exports[`HorizonCard render renders the horizon card if no error is present on data 1`] = `
<!---->
      <ha-card>
        <div class="horizon-card">
          HORIZON CARD CONTENT
        </div>
      </ha-card>
    
`;
